### DISCUSSION TITLE
Bug: `interpreter.state.can()` always be `false` if the interpreter started with an initial state from DB

### DISCUSSION THREAD

**Discussion by suerta-git** [statelyai/xstate#3094] [Q&A]
### Description

I am storing states in DB as JSON string following doc's recommendation. And restore them like:

```ts
const state = JSON.parse(jsonString)
const service = interpret(machine).start(state)
```

Then I want to use `state.can()` to test an event, but its result always be `false`

```ts
service.state.can(validEvent)  // false
```

If the interpreter started with a non-parsed state, the result is correct.

### Expected result

`true` for a valid event

### Actual result

`false`

### Reproduction

https://codesandbox.io/s/wizardly-perlman-01g01l?file=/src/index.ts

**Comment by davidkpiano** [comment]
Thanks for the bug report.

It is not recommended to read `service.state` directly, and to instead get that from `service.onTransition(state => {...})` or `service.subscribe(state => {...})`.

Since the state needs a reference to the machine for `state.can(...)` to work, a state (re)created _outside_ of the machine will never have the expected `state.can(...)` behavior. What would you expect here - an error, or a warning?

**Comment by suerta-git** [comment]
Thanks for your explanation. I am not sure what behavior is reasonable in this scenario.

Here I just want to check if an event is valid for current state before I call `service.send()` method, so I use `state.can()` here. 

Actual I prefer to get an error if I pass an invalid event to `service.send()`, but currently it won't.

**Reply by suerta-git** [reply]
@davidkpiano I find there are some notes about "strict mode":

> If neither a state nor any of its ancestor (parent) states handle an event, no transition happens. In strict mode (specified in the [machine configuration](https://xstate.js.org/docs/guides/machines.html#configuration)), this will throw an error.

But I don't find any detailed introductions about it in the referenced doc.

**Reply by suerta-git** [reply]
Looks like this is what I'm looking for. As I mentioned above:
> I prefer to get an error if I pass an invalid event to service.send()

**Reply by Andarist** [reply]
wouldn't it be better to be validated by a type system? without runtime schemas it's hard to define what a "valid event" actually is

**Reply by suerta-git** [reply]
Sorry, my description is not very accurate.

About the "valid event", I mean a event which is allowed at current state.

For example:

```ts
const machine = createMachine<ToggleContext, ToggleEvent>({
  id: "machine",
  initial: "inactive",
  context: {
    count: 0
  },
  states: {
    inactive: {
      on: { TOGGLE_ON: "active" }
    },
    active: {
      on: { TOGGLE_OFF: "inactive" }
    }
  }
});
```

When the machine is in `inactive` state, I call `TOGGLE_ON` a valid event and `TOGGLE_OFF` an invalid event.

**Reply by Andarist** [reply]
Note that if we consider big picture guarantees then this is not something that can be guaranteed. Things can get a little fuzzy because the machine might have some events already queued.

Just theoretically speaking now. If we put a "throw" at the beginning of the `service.send` then:
- it's possible that we won't throw, we queue the received event, process what is already in the queue and **then** try processing the event received by your `service.send` call. At that time, it's possible that the machine can already be in another state that would be, from your perspective, "invalid"
- and the opposite is true, we can "throw" but it's possible that the machine would "reach" the valid state after processing the currently queued events, so you would miss that event

The other option would be to throw when we actually pop the event from the queue - but that moment is outside of your control, so how do we notify you about a failure? and how do you even handle such an error?

**Reply by suerta-git** [reply]
Sorry I am new on xstate so not familiar with the mechanism of events processing.

General speaking, as an event driven system (regardless of side effects), its state at any time should be predictable and traceable. 
Even if there are events in queue that are not yet processed, it should be able to calculate the final state of the system, and be able to check the received event from `service.send`.

If it's impossible, I would like to know what is causing machine's state to be unpredictable. Because I'm trying to build a stateless system (which may be destroyed and restarted at any time), so certainty is important to me. Thank you.
